### PR TITLE
fork-tree: expose silent failures of is_descendent_of

### PR DESCRIPTION
**Pull request by andresilva**
This can lead to silent failures caused by client code (`is_descendent_of`) which can lead to consistency issues with the data stored in the tree.

### CODE REVIEW

File: utils/fork-tree/src/lib.rs [hunk 1]
```diff
@@ -610,18 +614,20 @@ impl<H, N, V> ForkTree<H, N, V> where
 		// descendent (in this case the node wasn't finalized earlier presumably
 		// because the predicate didn't pass).
 		let mut changed = false;
-		self.roots.retain(|root| {
-			let retain =
-				root.number > number && is_descendent_of(hash, &root.hash).unwrap_or(false) ||
-				root.number == number && root.hash == *hash ||
-				is_descendent_of(&root.hash, hash).unwrap_or(false);
+		let mut roots = Vec::new();
+		std::mem::swap(&mut roots, &mut self.roots);
```
Comment (bkchr) [inline utils/fork-tree/src/lib.rs:618]: You want mem::take

File: utils/fork-tree/src/lib.rs [hunk 2]
```diff
@@ -415,15 +415,16 @@ impl<H, N, V> ForkTree<H, N, V> where
 		// another fork not part of the tree). make sure to only keep roots that
 		// are part of the finalized branch
 		let mut changed = false;
-		self.roots.retain(|root| {
-			let retain = root.number > number && is_descendent_of(hash, &root.hash).unwrap_or(false);
+		let mut roots = Vec::new();
+		std::mem::swap(&mut roots, &mut self.roots);
```
Comment (bkchr) [inline utils/fork-tree/src/lib.rs:419]: Same

**Review by bkchr — COMMENTED** [review]
How often is this stuff called? Do we may need to allocate the new vector with a start capacity?

**Comment by andresilva** [comment]
@bkchr Updated. Normally these vecs will only have one or two elements unless we are forking really hard :P

**Review by octol — APPROVED** [review]

**Review by bkchr — APPROVED** [review]

**Pull request merged**